src/DynamoCore/Models/DynamoModel.cs

```diff
@@ -767,6 +767,28 @@ private void RegisterCustomNodeDefinitionWithEngine(CustomNodeDefinition definit
                 Workspaces.OfType<HomeWorkspaceModel>().SelectMany(ws => ws.Nodes),
                 definition,
                 DebugSettings.VerboseLogging);
+
+            MarkAllDependenciesAsModified(definition);
+        }
+
+        /// <summary>
+        /// Get all function instances or directly or indrectly dependo on the 
+        /// specified function definition and mark them as modified so that 
+        /// their values will be re-queryed.
+        /// </summary>
+        /// <param name="functionId"></param>
+        /// <returns></returns>
+        private void MarkAllDependenciesAsModified(CustomNodeDefinition def)
+        {
+            var homeWorkspace = Workspaces.OfType<HomeWorkspaceModel>().FirstOrDefault();
+            var dependencies = CustomNodeManager.GetAllDependenciesGuids(def);
+            var funcNodes = homeWorkspace.Nodes.OfType<Function>();
+            var dirtyNodes = funcNodes.Where(n => dependencies.Contains(n.Definition.FunctionId));
+
+            foreach (var node in dirtyNodes)
+            {
+                node.OnNodeModified();
```

Comment:
This causes `HomeWorkspaceModel.OnNodesModified` method to be called multiple times, causing `Run` to be called multiple times. Instead of doing this, we should collect them and do once in the `WorkspaceModel`. Say, let's expose a new method called `HomeWorkspaceModel.MarkNodesAsDirty(bool forceExecute)`, then call that method from here. This new method should have a similar implementation as the second part of `HomeWorkspaceModel.ResetEngine` method. Once you have this new method, don't forget to update `ResetEngine` to call it too.